WIP: Add jit_compile/1, very basic JIT1, skeleton of JIT2

aarroyoc opened the pull request:
I open this PR as I think it's good to have some code merged to not diverge a lot from the master branch.

All the JIT code is under the `jit` feature which is not enabled by default, so you shouldn't expect any change if you don't enable it (apart from an extra predicate, `jit_compile/1`, that does nothing).

There are two JITs right now:
- JIT1 was my first attempt. It tries to share all the data structures with the interpreter. I saw some problems with this approach so I stopped it.
- JIT2 is a new approach. I try to learn from the JIT1, it will manage more things on its own.

I also added a JIT CI test

UWN commented:
Just a naive question: Have you the possibility to abolish a definition (even if it is static)? The systems that do this correctly are rare (like 1, id est SICStus)

UWN commented:
Within ISO, there is only `abolish/1`to abolish dynamic predicates. SICStus offers the additional functionality to abolish also static predicates with [abolish/2](https://sicstus.sics.se/sicstus/docs/latest4/html/sicstus.html/mpg_002dref_002dabolish.html#index-abolish_002f_005b1_002c2_005d-_0028built_002din_002c-ref-page_0029-1) with option force(true). Being able to abolish static code on a predicate-by-predicate basis is a bit much to ask for since this prevents many optimizations. Maybe the granularity could be somewhat coarser. Like the entire Prolog text that was prepared for execution. Or some dependency mechanism. However, if code is abolished it must be conforming to the logical update semantics...

The most important use case is indeed to reload some code.

UWN commented:
> The most important use case is indeed to reload some code.

And the most important current use case of this would indeed be exhaustive testing. Currently, this does not work for `asserta/1` &amp; `rectact/1`and thus the WAM code generation is not much tested either. Even SICStus had (some time ago) issues there.

aarroyoc commented:
I updated the PR, I removed the JIT1, now all the focus is on JIT2, which can only do very basic predicates right now (like `a(1,2,3)` and query: `a(X, Y, Z)`. But it implements the skeleton of the new architecture, it ships a deref/store implementation in pure CLIF (Cranelift assembly). Everything it's still under a feature flag, so I think it's better to merge it, so code doesn't diverge too much between branches.

triska commented:
@aarroyoc: Thank you a lot for working on this extremely impressive feature!

In order to merge this with minimal chance for conflicts with other PRs and existing ongoing developments in other branches, do you think it would be good to reduce the changes to exactly those that should actually be kept, by rebasing the PR and removing features that were only present intermittently, if any?
